storage/innobase/include/trx0trx.h

```diff
@@ -902,6 +903,146 @@ struct TrxVersion {
 
 typedef std::list<TrxVersion, ut_allocator<TrxVersion> > hit_list_t;
 
+/** Utility class to collect and report slow query log InnoDB extension
+stats */
+class trx_stats {
+	public:
+	trx_stats() {
+	/* Always created in a zeroed memory block */
+#ifdef UNIV_DEBUG
+		ut_ad(lock_que_wait_ustarted == 0);
+		ut_ad(take_stats == false);
+#endif
+	}
+
+	/**
+	Register, if needed, start of an I/O read or wait.
+
+	@param	trx	transaction to account I/O to or NULL
+	@param	bytes	number of bytes to read or 0 if this is a wait for an
+	 already posted read in progress
+	@return value to be passed to end_io_read
+	*/
+	MY_NODISCARD
+	static ib_uint64_t
+	start_io_read(trx_t *trx, ulint bytes);
+
+	/**
+	Register start of an I/O read or wait.
+
+	@param	trx	transaction to account I/O to
+	@param	bytes	number of bytes to read or 0 if this is a wait for an
+	 already posted read in progress
+	@return value to be passed to end_io_read
+	*/
+	MY_NODISCARD
+	static ib_uint64_t
+	start_io_read(const trx_t &trx, ulint bytes);
+
+	/**
+	Register, if needed, end of an I/O read or wait.
+
+	@param	trx		transaction to account I/O to or NULL
+	@param	start_time	return value of start_io_read
+	*/
+	static void
+	end_io_read(trx_t *trx, ib_uint64_t start_time);
+
+	/**
+	Register end of an I/O read or wait.
+
+	@param	trx		transaction to account I/O to
+	@param	start_time	return value of start_io_read*/
+	static void
+	end_io_read(const trx_t &trx, ib_uint64_t start_time);
+
+	/**
+	Register, if needed, a single untimed I/O read.
+
+	@param	trx	transaction to account I/O to or NULL
+	@param	bytes	number of bytes read
+	*/
+	static void
+	bump_io_read(trx_t *trx, ulint bytes);
+
+	/**
+	Register a single untimed I/O read.
+
+	@param	trx	transaction to account I/O to
+	@param	bytes	number of bytes read
+	*/
+	static void
+	bump_io_read(const trx_t &trx, ulint bytes);
+
+	/**
+	Register, if needed, a start of lock wait */
+	void
+	start_lock_wait() {
```

Comment:
Any reason why only this method is defined here in place whereas others are defined externally via `inline` keyword?

Reply:
Also consider an auxiliary RAII class which would do `start_lock_wait()` in constructor and `stop_lock_wait()` in destructor.

Reply:
I'd prefer everything inplace but forward references happen. Thus it's a mix of the two. Do you want everything defined externally?

Reply:
I'd rather do RAII as a follow-up. Too many things going at once in the PR already

Reply:
Regarding `inline`: just wanted to hear that there is a reason for doing this. I doubt this is worth reworking.